res/controllers/Denon-MC4000-scripts.js

```diff
@@ -0,0 +1,544 @@
+    }
+    // Initalization of controller states
+    // TODO: find equivalent message for MC4000
+    // tell controller to send MIDI messages with positions of faders and knobs
+    // midi.sendShortMsg(0xB0, 0x7F, 0x7F);
+};
+
+MC4000.shutdown = function () {};
+
+MC4000.padColors = {
+    off: 0,
+    red: 2,
+    blue: 1,
+    purple: 3
+};
+
+// Found by averaging total 'ticks' received over 100 rotations
+MC4000.jogWheelTicksPerRevolution = 605;
+
+MC4000.scratchParams = {
+    recordSpeed: 33.3333333333,
+    alpha: (1.0/8),
+    beta: (1.0/8)/32.0
+};
+
+MC4000.jogParams = {
+    sensitivity: 25,
+    maxJogValue: 3
+};
+
+MC4000.rateRanges = [0.08, 0.16, 0.5];
+
+MC4000.OnVuMeterChangeL = function(value, group, control) {
+    midi.sendShortMsg(0xBF, 0x44, value*0x7F);
+}
+
+MC4000.OnVuMeterChangeR = function(value, group, control) {
+    midi.sendShortMsg(0xBF, 0x45, value*0x7F);
+}
+
+MC4000.Deck = function (channel) {
+    // Some state variables
+    this.isVinylMode = false;
+    this.rateMsbValue = 0;
+    this.midiChannel = channel;
+    this.engineChannel = channel + 1;
+    this.group = '[Channel' + this.engineChannel + ']';
+    this.currentRateIndex = 0;
+    
+    // Set default rate range (to +/- 8% like Serato and Technics)
+    engine.setValue(this.group, "rateRange", MC4000.rateRanges[this.currentRateIndex]);
+    // Initialize vinyl mode LED
+    midi.sendShortMsg(0x90 + channel, 0x07, this.isVinylMode ? 0x7F: 0x00);
+    // Callback for touching the jog wheel platter
+    this.platterTouch = function (channel, control, value, status, group) {
+        if (this.isVinylMode) {
+            if (value > 0) {
+                engine.scratchEnable(this.engineChannel, MC4000.jogWheelTicksPerRevolution, 
+                    MC4000.scratchParams.recordSpeed , MC4000.scratchParams.alpha, MC4000.scratchParams.beta);
+            } else { 
+                engine.scratchDisable(this.engineChannel);
+            }
+        }
+    }
+    // Callback for toggling the vinyl mode button
+    this.vinylModeToggle = function (channel, control, value, status, group) {
+        if (value > 0) {
+            this.isVinylMode = !this.isVinylMode;
+            midi.sendShortMsg(0x90 + channel, 0x07, this.isVinylMode ? 0x7F: 0x00);
+        }
+    }
+    // Key lock toggle button
+    this.keylock = new ToggleButton([0x90 + channel, 0x0D], this.group, 'keylock');
+    // Rate range toggle button callback
+    this.rateRange = function(midichan, control, value, status, group) {
+        if (value == 0) {
+            return; // don't respond to note off messages
+        }
+        this.currentRateIndex++;
+        if (this.currentRateIndex >= MC4000.rateRanges.length) {
+            this.currentRateIndex = 0;
+        }
+        engine.setValue(this.group, "rateRange", MC4000.rateRanges[this.currentRateIndex]);
+    };
+    // Sync toggle button
+    // TODO: shift + sync should be used to diable sync instead of using toggle
```

Comment:
I think the way it works in Serato is that when you press the sync button while the sync function is enabled, the sync is recalculated at the current time point. You need to use shift + sync to turn off the sync function entirely. I'm not sure how or if I should map that same functionality to Mixxx.